longhaultests/src/iothub_longhaul.ts

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env node
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+import * as uuid from 'uuid';
+import * as async from 'async';
+import * as dbg from 'debug';
+const debug = dbg('longhaul:main');
+
+import { Registry, ConnectionString as HubConnectionString } from 'azure-iothub';
+import { ConnectionString as DeviceConnectionString } from 'azure-iot-device';
+
+import { D2CSender } from './d2c_sender';
+
+const MAX_EXECUTION_TIME = parseInt(process.env.MAX_EXECUTION_TIME_SECONDS) * 1000;
+const SEND_INTERVAL = parseInt(process.env.DEVICE_SEND_INTERVAL_SECONDS) * 1000;
+const SEND_TIMEOUT = parseInt(process.env.DEVICE_SEND_TIMEOUT_SECONDS) * 1000;
+
+const MAX_CREATE_TIME = 30000;
+const MAX_DELETE_TIME = 30000;
+const MAX_START_TIME = 30000;
+const MAX_STOP_TIME = 30000;
+
+const ERROR_EXIT_CODE = -1;
+const OK_EXIT_CODE = 0;
+
+const hubConnectionString = process.env.IOTHUB_CONNECTION_STRING;
+const registry = Registry.fromConnectionString(hubConnectionString);
+
+const deviceId = 'node-longhaul-sak-' + uuid.v4();
+let protocol;
+
+/* tslint:disable:no-var-requires */
+switch (process.env.DEVICE_PROTOCOL) {
+  case 'amqp':
+    protocol = require('azure-iot-device-amqp').Amqp;
+  break;
+  case 'amqp-ws':
+    protocol = require('azure-iot-device-amqp').AmqpWs;
+  break;
+  case 'mqtt':
+    protocol = require('azure-iot-device-mqtt').Mqtt;
+  break;
+  case 'mqtt-ws':
+    protocol = require('azure-iot-device-mqtt').MqttWs;
+  break;
+  case 'http':
+    protocol = require('azure-iot-device-mqtt').Http;
+  break;
+  default:
+    debug('unknown protocol: ' + process.env.DEVICE_PROTOCOL);
+    process.exit(ERROR_EXIT_CODE);
+}
+/* tslint:enable:no-var-requires */
+
+const createDevice = (callback) => {
+  debug('creating device: ' + deviceId);
+  async.timeout(registry.create.bind(registry), MAX_CREATE_TIME)({ deviceId: deviceId }, (err, deviceInfo) => {
```

Comment:
>bind(registry) [](start = 32, length = 14)

E V I L!

Reply:
I'm trying to wrap my head around the necessity of the bind call here, and I'm also wondering if it would be cleaner to create all the timeout wrappers together in a separate function.  I'm thinking maybe build an entire wrapped object, like promisify does, only with async.timeout()  It wouldn't have to use forEach and wrap every method -- it could just be hardcoded functions/timeout values.  I just like the idea of having an object that looks exactly like the registry object, with the additional timeout functionality wrapping all the calls.    (side thought: would this make sense with retry?  Maybe too late now for the node SDK, but could be a useful paradigm)

---
In reply to: [177508442](https://github.com/Azure/azure-iot-sdk-node/pull/257#discussion_r177508442) [](ancestors = 177508442)

Reply:
definitely having the same thoughts on the side. I'll refactor as we "Generalize" this pattern to more features. right now i want to get it out there and get builds to run and succeed.

---
In reply to: [177514489](https://github.com/Azure/azure-iot-sdk-node/pull/257#discussion_r177514489) [](ancestors = 177514489,177508442)